### PR TITLE
Saga tutorial lesson 2

### PR DESCRIPTION
**Pull request by mauroservienti**

### CODE REVIEW

File: tutorials/nservicebus-sagas/2-timeouts/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+Consider what happens when the buyer's remorse period has ended. The saga has been marked complete but maybe the Cancel button still appears on the user's screen and they click it. Assuming a `CancelOrder` command still gets fired, nothing will happen. The saga instance is already complete so there is nothing left to process this command. In effect, we can't cancel an order that has already been placed. Similarly, we can't complete an order that has already been processed. `MarkAsComplete` handles both of these scenarios for us.
+
+NOTE: Technically, it's not true that _nothing_ will happen when a saga message is received after the instance has completed. We can implement the `IHandleSagaNotFound` interface to [perform logic](https://docs.particular.net/nservicebus/sagas/saga-not-found) in these scenarios. For example, we could inform the customer that the buyer's remorse period has ended so the order can no longer be cancelled.
```
Comment (ramonsmits) [inline tutorials/nservicebus-sagas/2-timeouts/tutorial.md]:
```suggestion
NOTE: Technically, it's not true that _nothing_ will happen when a saga message is received after the instance has completed. We can implement the `IHandleSagaNotFound` interface to [perform logic](/nservicebus/sagas/saga-not-found.md) in these scenarios. For example, we could inform the customer that the buyer's remorse period has ended so the order can no longer be cancelled.
```

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyStartedByPlaceOrder/BuyersRemorsePolicyStartedByPlaceOrder.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        : Saga<BuyersRemorseState>
+        , IAmStartedByMessages<PlaceOrder>
+    {
+        static ILog log = LogManager.GetLogger<BuyersRemorsePolicy>();
+
+        public Task Handle(PlaceOrder message, IMessageHandlerContext context)
+        {
+            log.Info($"Received PlaceOrder, OrderId = {message.OrderId}");
+
+            Data.CustomerId = message.CustomerId;
```
Comment (ramonsmits) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyStartedByPlaceOrder/BuyersRemorsePolicyStartedByPlaceOrder.cs]:
```suggestion
            Data.OrderId = message.OrderId;
```

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutRequest/BuyersRemorseTimeoutRequest.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            
+        }
+
+        #region BuyersRemorseTimeoutRequest
+
+        public async Task Handle(PlaceOrder message, IMessageHandlerContext context)
+        {
+            log.Info($"Received PlaceOrder, OrderId = {message.OrderId}");
+            Data.CustomerId = message.CustomerId;
```
Comment (ramonsmits) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutRequest/BuyersRemorseTimeoutRequest.cs]:
```suggestion
            Data.OrderId = message.OrderId;
```

File: tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    internal class BuyersRemorseIsOver
+    {
+
+    }
+
+    public class BuyersRemorseState : ContainSagaData
+    {
+        public string OrderId { get; set; }
+        public string CustomerId { get; set; }
```
Comment (ramonsmits) [inline tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs]:
```suggestion
```

**Comment by mauroservienti** [comment]
Pushed a change to fix the build failure and to improve the last bit (changes to CientUI) of the tutorial.

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancellingOrders/BuyersRemorseCancellingOrders.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                var key = Console.ReadKey();
+                Console.WriteLine();
+
+                switch (key.Key)
+                {
+                    case ConsoleKey.P:
+                        // Instantiate the command
+                        var command = new PlaceOrder
+                        {
+                            OrderId = Guid.NewGuid().ToString()
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancellingOrders/BuyersRemorseCancellingOrders.cs]: Never understood why we ToString those :-)

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyMapping/BuyersRemorsePolicyMapping.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+﻿namespace Core_7.BuyersRemorsePolicyMapping
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System.Threading.Tasks;
+
+    #region BuyersRemorsePolicyMapping
+
+    class BuyersRemorsePolicy
+        : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyMapping/BuyersRemorsePolicyMapping.cs]: We aren't consistent. This line should be appended to previous one, like the saga earlier.

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyMapping/BuyersRemorsePolicyMapping.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+﻿namespace Core_7.BuyersRemorsePolicyMapping
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System.Threading.Tasks;
+
+    #region BuyersRemorsePolicyMapping
+
+    class BuyersRemorsePolicy
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyMapping/BuyersRemorsePolicyMapping.cs]:
```suggestion
    class BuyersRemorsePolicy : Saga<BuyersRemorseState>
```

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyStartedByPlaceOrder/BuyersRemorsePolicyStartedByPlaceOrder.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
+﻿namespace Core_7.BuyersRemorsePolicyStartedByPlaceOrder
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System.Threading.Tasks;
+
+    #region BuyersRemorsePolicyStartedByPlaceOrder
+
+    class BuyersRemorsePolicy
+        : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorsePolicyStartedByPlaceOrder/BuyersRemorsePolicyStartedByPlaceOrder.cs]: Same here

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutClassDefinition/BuyersRemorseTimeoutClassDefinition.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
+﻿namespace Core_7.BuyersRemorseTimeoutClassDefinition
+{
+    #region BuyersRemorseTimeoutClassDefinition
+
+    class BuyersRemorsePolicy
+        : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutClassDefinition/BuyersRemorseTimeoutClassDefinition.cs]: And here... I see a pattern, but I dislike the breaking up continuously. I prefer the first thing to be on the same line.

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutHandling/BuyersRemorseTimeoutHandling.cs [hunk 10]
```diff
@@ -0,0 +1,56 @@
+﻿namespace Core_7.BuyersRemoreseTimeoutHandling
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System.Threading.Tasks;
+
+    #region BuyersRemorseTimeoutHandling
+
+    class BuyersRemorsePolicy
+        : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutHandling/BuyersRemorseTimeoutHandling.cs]: And here :-)

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancelOrderHandling/BuyersRemorseCancelOrderHandling.cs [hunk 11]
```diff
@@ -0,0 +1,98 @@
+﻿namespace Core_7.BuyersRemorseCancelOrderHandling
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System;
+    using System.Threading.Tasks;
+
+    #region BuyersRemorseCancelOrderHandling
+
+    class BuyersRemorsePolicy : Saga<BuyersRemorseState>,
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancelOrderHandling/BuyersRemorseCancelOrderHandling.cs]: For reference later on, the saga is on the same line. Has my preference, but every other occassion it's the other way around. Again, I prefer this version.

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutRequest/BuyersRemorseTimeoutRequest.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
+﻿namespace Core_7.BuyersRemorseTimeoutRequest
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System;
+    using System.Threading.Tasks;
+
+    class BuyersRemorsePolicy : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseTimeoutRequest/BuyersRemorseTimeoutRequest.cs]: Yay! Here is how I like to see it. Unfortunately not inside a region, so no snippet ;-)

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/EmptyBuyersRemorsePolicy/EmptyBuyersRemorsePolicy.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
+﻿namespace Core_7.EmptyBuyersRemorsePolicy
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+
+#pragma warning disable 1998
+
+    #region EmptyBuyersRemorsePolicy
+    class BuyersRemorsePolicy : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/EmptyBuyersRemorsePolicy/EmptyBuyersRemorsePolicy.cs]: And here again on the same line, within a region.

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Snippets.sln.DotSettings [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Snippets/Snippets.sln.DotSettings]: This entire file needs to be ignored and removed, not?

File: tutorials/nservicebus-sagas/2-timeouts/Solution/RetailDemo.sln.DotSettings [hunk 15]
```diff
@@ -0,0 +1,6 @@
+<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Solution/RetailDemo.sln.DotSettings]: This file needs to be removed not?

File: tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs [hunk 16]
```diff
@@ -0,0 +1,66 @@
+﻿namespace Core_7.BuyersRemorseTimeoutRequest
+{
+    using NServiceBus;
+    using NServiceBus.Logging;
+    using System;
+    using System.Threading.Tasks;
+    using Messages;
+
+    class BuyersRemorsePolicy
+        : Saga<BuyersRemorseState>
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs]: Again, additional line.

File: tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+            await context.Publish(orderPlaced);
+
+            MarkAsComplete();
+        }
+
+        public Task Handle(CancelOrder message, IMessageHandlerContext context)
+        {
+            log.Info($"Order #{message.OrderId} was cancelled.");
+
+            //TODO: Update status in database?
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs]:
```suggestion
            //TODO: Possibly publish an OrderCancelled event?
```

File: tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        public async Task Timeout(BuyersRemorseIsOver state, IMessageHandlerContext context)
+        {
+            var orderPlaced = new OrderPlaced
+            {
+                OrderId = Data.OrderId
+            };
+
+            // TODO: Save order state in database?
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs]: I don't get this. The saga is already stored inside the database? What do we need to do? Does `OrderPlaced` not do everything we'd need? People might think we need to actually access our business database and store something?

File: tutorials/nservicebus-sagas/2-timeouts/Solution/Shipping/ShippingPolicy.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
+﻿using Messages;
+using NServiceBus;
+using NServiceBus.Logging;
+using System.Threading.Tasks;
+
+namespace Shipping
+{
+    class ShippingPolicy : Saga<ShippingPolicyData>,
```
Comment (dvdstelt) [inline tutorials/nservicebus-sagas/2-timeouts/Solution/Shipping/ShippingPolicy.cs]: :cheer:

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancellingOrders/BuyersRemorseCancellingOrders.cs (hunk 5 above)
Comment (mauroservienti) [reply tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancellingOrders/BuyersRemorseCancellingOrders.cs]: Me neither, but to change it we need to change lesson 1 as well. That is out of scope.

File: tutorials/nservicebus-sagas/2-timeouts/Solution/RetailDemo.sln.DotSettings (hunk 15 above)
Comment (mauroservienti) [reply tutorials/nservicebus-sagas/2-timeouts/Solution/RetailDemo.sln.DotSettings]: Good question, it's also in lesson 1. I don't have R# so I don't need it, I have no idea why it's there.

File: tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs (hunk 16 above)
Comment (mauroservienti) [reply tutorials/nservicebus-sagas/2-timeouts/Solution/Sales/BuyersRemorsePolicy.cs]: Even if I agree this is in the final solution, I'd rather keep it consistent with lesson 1. Thoughts?

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Snippets.sln.DotSettings (hunk 14 above)
Comment (mauroservienti) [reply tutorials/nservicebus-sagas/2-timeouts/Snippets/Snippets.sln.DotSettings]: Good question, it's also in lesson 1. I don't have R# so I don't need it, I have no idea why it's there.

File: tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancelOrderHandling/BuyersRemorseCancelOrderHandling.cs (hunk 11 above)
Comment (ramonsmits) [reply tutorials/nservicebus-sagas/2-timeouts/Snippets/Core_7/BuyersRemorseCancelOrderHandling/BuyersRemorseCancelOrderHandling.cs]: Well, inheritance on the next line seems like the standard for messages too etc. I'm not against using a single line but it should be consistent on all types.

What do you think @mauroservienti ?

**Pull request merged**